wgpu-core/src/device/queue.rs

```diff
@@ -28,16 +28,29 @@ use thiserror::Error;
 /// without a concrete moment of when it can be cleared.
 const WRITE_COMMAND_BUFFERS_PER_POOL: usize = 64;
 
-pub type OnSubmittedWorkDoneCallback = unsafe extern "C" fn(user_data: *mut u8);
 #[repr(C)]
-#[derive(Clone, Copy, Debug)]
-pub struct SubmittedWorkDoneClosure {
-    pub callback: OnSubmittedWorkDoneCallback,
-    pub user_data: *mut u8,
+pub struct SubmittedWorkDoneClosureC {
+    callback: unsafe extern "C" fn(user_data: *mut u8),
+    user_data: *mut u8,
 }
 
-unsafe impl Send for SubmittedWorkDoneClosure {}
-unsafe impl Sync for SubmittedWorkDoneClosure {}
+unsafe impl Send for SubmittedWorkDoneClosureC {}
+
+pub enum SubmittedWorkDoneClosure {
+    Rust { callback: Box<dyn FnOnce() + Send> },
+    C { inner: SubmittedWorkDoneClosureC },
+}
```

Comment:
Any function that accepts this type (for example, `Global::on_submitted_work_done`) needs to be marked `unsafe`, since there are conditions on its safe use (namely, that the function pointer had better be good) that Rust's type system cannot verify. See the comments on `BufferMapCallback` for an alternative suggestion that I think is a bit cleaner.